File: day02/Day02.hs
shape [_,"X"] = 1
shape [_,"Y"] = 2
shape [_,"Z"] = 3

outcome ["C","X"] = 6
outcome ["A","Y"] = 6
outcome ["B","Z"] = 6
outcome ["A","X"] = 3
outcome ["B","Y"] = 3
outcome ["C","Z"] = 3
outcome _ = 0

outcome2 [_,"X"] = 0
outcome2 [_,"Y"] = 3
outcome2 [_,"Z"] = 6

shape2 ["A","X"] = 3
shape2 ["A","Y"] = 1
shape2 ["A","Z"] = 2
shape2 ["B","X"] = 1
shape2 ["B","Y"] = 2
shape2 ["B","Z"] = 3
shape2 ["C","X"] = 2
shape2 ["C","Y"] = 3
shape2 ["C","Z"] = 1

main = do input <- map words <$> lines <$> readFile "input.txt"
          print $ sum $ map shape input ++ map outcome input
          print $ sum $ map shape2 input ++ map outcome2 input

File: day02/day02.rs
use std::fs::File;
use std::io::Read;

fn anyn(noob: &str, n: usize) -> bool {
    "abcdefghijklmnopqrstuvwxyz".chars().any(|x| noob.chars().filter(|c| *c == x).count() == n)
}

fn countn(input: &String, n: usize) -> usize {
    input.lines().filter(|line| anyn(line, n)).count()
}

fn main() {
    let mut file = File::open("input.txt").unwrap();
    let mut input = String::new();
    file.read_to_string(&mut input).unwrap();
    // part 1
    println!("{}", countn(&input, 2) * countn(&input, 3));
    // part 2
    for (i, x) in input.lines().enumerate() {
        for y in input.lines().skip(i) {
            let z: String = x.chars().zip(y.chars()).filter_map(|(x,y)| if x == y { Some(x) } else { None }).collect();
            if z.len() == x.len() - 1 {
                println!("{}", z);
            }
        }
    }
}
